util/chplenv/compile_link_args_utils.py

```diff
@@ -0,0 +1,198 @@
+
+    if mem == "jemalloc":
+        # set -DCHPL_JEMALLOC_PREFIX=chpl_je_
+        # this is needed since it affects code inside of headers
+        bundled.append("-DCHPL_JEMALLOC_PREFIX=chpl_je_")
+
+    return (bundled, system)
+
+# Returns the runtime -L and -l args according
+# to the current configuration, for a target (not host) compile.
+# Returns tuple of (bundled_args, system_args)
+def get_runtime_link_args(runtime_subdir):
+    bundled = [ ]
+    system = [ ]
+
+    lib = chpl_home_utils.get_chpl_runtime_lib()
+    locale_model = chpl_locale_model.get()
+
+    bundled.append("-L" + os.path.join(lib, runtime_subdir))
+    bundled.append("-lchpl")
+
+    if locale_model == "gpu":
+        # If compiling for GPU locales, add CUDA to link path,
+        # and add cuda libraries
+        cuda_path = chpl_gpu.get_cuda_path()
+        system.append("-L", os.path.join(cuda_path, "lib64"))
+        system.append("-lcuda")
+        system.append("-lcudart")
+
+    # always link with the math library
+    system.append("-lm")
+    # always link with the pthreads library
+    system.append("-lpthread")
+
+    return (bundled, system)
+
+# Returns a dictionary containing keys
+#   'host_compile'
+#   'host_link'
+#   'target_compile'
+#   'target_link'
+#
+# where each value is a tuple of compile/link args
+#  (bundled_args, system_args)
+def compute_internal_compile_link_args(runtime_subdir):
+    # compute the compiler / link args
+    # each of these is bundled, system
+    host_compile = ([ ], [ ])
+    host_link = ([ ], [ ])
+    tgt_compile = ([ ], [ ])
+    tgt_link = ([ ], [ ])
+
+    skip_host = os.environ.get('CHPLENV_SKIP_HOST', None)
+
+    # add runtime includes and defines
+    extend2(tgt_compile, get_runtime_includes_and_defines())
+    extend2(tgt_link, get_runtime_link_args(runtime_subdir))
+
+    # add 3p arguments
+
+    if (chpl_llvm.get() == 'bundled' or
+        chpl_llvm.get() == 'system'):
+        if not skip_host:
+            extend2(host_compile, chpl_llvm.get_host_compile_args())
+            extend2(host_link, chpl_llvm.get_host_link_args())
+
+    extend2(tgt_compile, chpl_gmp.get_compile_args())
+    extend2(tgt_link, chpl_gmp.get_link_args())
+
+    extend2(tgt_compile, chpl_hwloc.get_compile_args())
+    extend2(tgt_link, chpl_hwloc.get_link_args())
+
+    if chpl_comm.get() == 'ofi':
+        extend2(tgt_compile, chpl_libfabric.get_compile_args())
+        extend2(tgt_link, chpl_libfabric.get_link_args())
+    elif chpl_comm.get() == 'gasnet':
+        extend2(tgt_compile, chpl_gasnet.get_compile_args())
+        extend2(tgt_link, chpl_gasnet.get_link_args())
+    elif chpl_comm.get() == 'ugni':
+        # If there isn't a hugepage module loaded, we need to request
+        # libhugetlbfs ourselves.
+        pe_product_list = os.environ.get('PE_PRODUCT_LIST', None)
+        if pe_product_list and 'HUGETLB' in pe_product_list:
+            tgt_link[1].append('-lhugetlbfs')
```

Comment:
Just as a general note, there's something I don't like about centralizing all this logic. I liked that logic specific to ugni before was in a ugni directory/file before. I think this is just the way it has to be in the new world, and it doesn't make sense to me to have a "third-party" ugni chplenv file just for this logic, but just wanted to mention it

Reply:
IMO it would be just fine to make `chpl_ugni.py` and put this in there. I did not do that here for various reasons. Would you like me to add a TODO comment that we might want to do that?

Reply:
I'm OK with the ugni logic here as is, and maybe splitting the logic off into a separate file or function down the road if we find it helpful.